File: vscode/test/rule-evaluator.test.js
const assert = require('node:assert/strict');
const test = require('node:test');

const { evaluateRules } = require('../out/core/RuleEvaluator');
const { ImeAction } = require('../out/core/types');

const defaultRules = {
  chineseBeforeRegex: '.*[\\u4e00-\\u9fa5]$',
  chineseAfterRegex: '^[\\u4e00-\\u9fa5].*',
  capsBeforeRegex: '.*[A-Z]{2,}[0-9_]?$',
  capsAfterRegex: '',
};

test('returns chinese when before text matches chinese rule', () => {
  assert.equal(evaluateRules('输入', '', defaultRules), ImeAction.CHINESE);
});

test('ignores punctuation next to caret when matching chinese context', () => {
  assert.equal(evaluateRules('输入，。！？', '', defaultRules), ImeAction.CHINESE);
});

test('ignores punctuation next to caret when matching text after caret', () => {
  assert.equal(evaluateRules('', '（《中文', defaultRules), ImeAction.CHINESE);
});

test('prefers matching context before caret when both sides match', () => {
  assert.equal(
    evaluateRules(
      'ABC',
      '中文',
      defaultRules,
    ),
    ImeAction.CAPS,
  );
});

test('returns english for letters and english punctuation before caret', () => {
  for (const before of ['a', ',']) {
    assert.equal(evaluateRules(before, '中文', defaultRules), ImeAction.ENGLISH);
  }
});

test('ignores numbers for chinese caps and english matching', () => {
  for (const [before, expected] of [
    ['中文1', ImeAction.CHINESE],
    ['HTTP1', ImeAction.CAPS],
    ['a1', ImeAction.ENGLISH],
  ]) {
    assert.equal(evaluateRules(before, '', defaultRules), expected);
  }
});

test('number-only left context does not override right match', () => {
  assert.equal(evaluateRules('1', '中文', defaultRules), ImeAction.CHINESE);
});

test('space-only left context does not override right match', () => {
  assert.equal(evaluateRules(' ', '中文', defaultRules), ImeAction.CHINESE);
});

test('ignores spaces and numbers for all matching modes', () => {
  for (const [before, expected] of [
    ['中文 1', ImeAction.CHINESE],
    ['HTTP 1', ImeAction.CAPS],
    ['a 1', ImeAction.ENGLISH],
  ]) {
    assert.equal(evaluateRules(before, '', defaultRules), expected);
  }
});

test('does not treat full-width punctuation as english', () => {
  assert.equal(evaluateRules('，', '中文', defaultRules), ImeAction.CHINESE);
});

test('treats english punctuation as english instead of chinese context', () => {
  assert.equal(evaluateRules('中文,', '', defaultRules), ImeAction.ENGLISH);
});

test('treats english punctuation after caret as english', () => {
  assert.equal(evaluateRules('', ',中文', defaultRules), ImeAction.ENGLISH);
});

test('returns caps when caps rule matches and chinese does not', () => {
  assert.equal(evaluateRules('HTTP_', '', defaultRules), ImeAction.CAPS);
});

test('does not ignore general punctuation when matching caps context', () => {
  assert.equal(evaluateRules('HTTP_,', '', defaultRules), ImeAction.ENGLISH);
});

test('ignores only hyphens and underscores when matching caps context', () => {
  assert.equal(evaluateRules('HTTP_-', '', defaultRules), ImeAction.CAPS);
});

test('returns english when rules are blank or invalid', () => {
  assert.equal(
    evaluateRules(
      'plain',
      'text',
      {
        chineseBeforeRegex: '',
        chineseAfterRegex: '[',
        capsBeforeRegex: '',
        capsAfterRegex: '[',
      },
    ),
    ImeAction.ENGLISH,
  );
});

File: core/src/test/kotlin/com/auto_switch_ime/core/rules/RuleEvaluatorTest.kt
package com.auto_switch_ime.core.rules

import com.auto_switch_ime.core.ImeAction
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class RuleEvaluatorTest {

    @Test
    fun `returns chinese when before text matches chinese rule`() {
        val action = RuleEvaluator.evaluate(
            before = "输入",
            after = "",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.CHINESE, action)
    }

    @Test
    fun `ignores punctuation next to caret when matching chinese context`() {
        val action = RuleEvaluator.evaluate(
            before = "输入，。！？",
            after = "",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.CHINESE, action)
    }

    @Test
    fun `ignores punctuation next to caret when matching text after caret`() {
        val action = RuleEvaluator.evaluate(
            before = "",
            after = "（《中文",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.CHINESE, action)
    }

    @Test
    fun `prefers matching context before caret when both sides match`() {
        val action = RuleEvaluator.evaluate(
            before = "ABC",
            after = "中文",
            chineseBeforeRegex = ".*[一-龥].*",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.CAPS, action)
    }

    @Test
    fun `returns english for letters and english punctuation before caret`() {
        for (before in listOf("a", ",")) {
            val action = RuleEvaluator.evaluate(
                before = before,
                after = "中文",
                chineseBeforeRegex = ".*[一-龥]$",
                chineseAfterRegex = "^[一-龥].*",
                capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
                capsAfterRegex = ""
            )

            assertEquals(ImeAction.ENGLISH, action)
        }
    }

    @Test
    fun `ignores numbers for chinese caps and english matching`() {
        val contexts = listOf(
            "中文1" to ImeAction.CHINESE,
            "HTTP1" to ImeAction.CAPS,
            "a1" to ImeAction.ENGLISH
        )

        for ((before, expected) in contexts) {
            val action = RuleEvaluator.evaluate(
                before = before,
                after = "",
                chineseBeforeRegex = ".*[一-龥]$",
                chineseAfterRegex = "^[一-龥].*",
                capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
                capsAfterRegex = ""
            )

            assertEquals(expected, action)
        }
    }

    @Test
    fun `number-only left context does not override right match`() {
        val action = RuleEvaluator.evaluate(
            before = "1",
            after = "中文",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.CHINESE, action)
    }

    @Test
    fun `space-only left context does not override right match`() {
        val action = RuleEvaluator.evaluate(
            before = " ",
            after = "中文",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.CHINESE, action)
    }

    @Test
    fun `ignores spaces and numbers for all matching modes`() {
        val contexts = listOf(
            "中文 1" to ImeAction.CHINESE,
            "HTTP 1" to ImeAction.CAPS,
            "a 1" to ImeAction.ENGLISH
        )

        for ((before, expected) in contexts) {
            val action = RuleEvaluator.evaluate(
                before = before,
                after = "",
                chineseBeforeRegex = ".*[一-龥]$",
                chineseAfterRegex = "^[一-龥].*",
                capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
                capsAfterRegex = ""
            )

            assertEquals(expected, action)
        }
    }

    @Test
    fun `does not treat full-width punctuation as english`() {
        val action = RuleEvaluator.evaluate(
            before = "，",
            after = "中文",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.CHINESE, action)
    }

    @Test
    fun `treats english punctuation as english instead of chinese context`() {
        val action = RuleEvaluator.evaluate(
            before = "中文,",
            after = "",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.ENGLISH, action)
    }

    @Test
    fun `treats english punctuation after caret as english`() {
        val action = RuleEvaluator.evaluate(
            before = "",
            after = ",中文",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.ENGLISH, action)
    }

    @Test
    fun `returns caps when caps rule matches and chinese does not`() {
        val action = RuleEvaluator.evaluate(
            before = "HTTP_",
            after = "",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.CAPS, action)
    }

    @Test
    fun `does not ignore general punctuation when matching caps context`() {
        val action = RuleEvaluator.evaluate(
            before = "HTTP_,",
            after = "",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.ENGLISH, action)
    }

    @Test
    fun `ignores only hyphens and underscores when matching caps context`() {
        val action = RuleEvaluator.evaluate(
            before = "HTTP_-",
            after = "",
            chineseBeforeRegex = ".*[一-龥]$",
            chineseAfterRegex = "^[一-龥].*",
            capsBeforeRegex = ".*[A-Z]{2,}[0-9_]?$",
            capsAfterRegex = ""
        )

        assertEquals(ImeAction.CAPS, action)
    }

    @Test
    fun `returns english when rules are blank or invalid`() {
        val action = RuleEvaluator.evaluate(
            before = "plain",
            after = "text",
            chineseBeforeRegex = "",
            chineseAfterRegex = "[",
            capsBeforeRegex = "",
            capsAfterRegex = "["
        )

        assertEquals(ImeAction.ENGLISH, action)
    }
}
